File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int kthSmallest(TreeNode root, int k) {
        int[] count = new int[2];
        count[0] = k;
        dfs(root, count);
        return count[1];
    }

    public void dfs(TreeNode root, int[] count) {
        if(root == null) return;

        dfs(root.left, count);
        count[0]--;
        if(count[0] == 0) {
            count[1] = root.val;
            return;
        } 
        dfs(root.right, count);
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if root is None:
            return True
        left_h = self.dfs(root.left) 
        right_h = self.dfs(root.right)

        if abs(left_h - right_h) > 1:
            return False

        return self.isBalanced(root.left) and self.isBalanced(root.right)

    def dfs(self, root) -> int:
        if root is None:
            return 0
        
        left_h = self.dfs(root.left)
        right_h = self.dfs(root.right)

        return max(left_h, right_h) + 1